doc/specs/#6899 - Action IDs/#6899 - Action IDs.md

```diff
@@ -0,0 +1,188 @@
+---
+author: Mike Griese @zadjii-msft
+created on: 2020-07-13
+last updated: 2020-07-13
+issue id: 6899
+---
+
+# Action IDs
+
+## Abstract
+
+This document is intended to serve as an addition to the [Command Palette Spec],
+as well as the [New Tab Menu Customization Spec].
+
+As we come to rely more on actions being a mechanism by which the user defines
+"do something in the Terminal", we'll want to make it even easier for users to
+re-use the actions that they've already defined, as to reduce duplicated json as
+much as possible. This spec proposes a mechanism by which actions could be
+uniquely identifiable, so that the user could refer to bindings in other
+contexts without needing to replicate an entire json blob.
+
+## Solution Design
+
+This spec was largely inspired by the following diagram from @DHowett:
+
+![figure 1](data-mockup.png)
+
+The goal is to introduce an `id` parameter by which actions could be uniquely
+refered to. We'd maintain a large list of default actions, each with unique
+`id`s set. The default bindings then would also point to actions defined in that
+default actions list, through their set `id`. Furthermore, as we add additional
+menus to the Terminal, like the customization for the new tab dropdown, or the
+tab context menu, or the `TermControl` context menu, they could all refer to
+these actions by `id`, rather than duplicating the same json.
+
+A cleaner version of the above mockup might look like:
+![figure 2](data-mockup-002.png)
+
+We'll first parse `actions` into a map of `id->Action`s. These ids must be
+unique. When layering `actions`, if a later settings file contains an action
+with the same `id`, it will replace the current value. In this way, users can
+redefine actions, or remove default ones (with something like `{ "id":
+"Terminal.OpenTab", "command":null }`
+
+> 👉 NOTE: The IDs for the default actions will need to be manually created, not
+> autogenerated. Every action in the default `bindings` should probably be given
+> an `id` and defined in `actions`.
+
+### Existing Scenarios
+
+When parsing `bindings`, we'll check for the presence of an `id` key. If that's
+found, we'll use the action with that global ID as the action for this binding.
+If there's no `id` set, then we'll continue parsing as we always have. The user
+doesn't _need_ to define any actions in `actions` - they could solely define
+actions in the bindings.
+
+**TODO**: When parsing a binding with both an `id` and a `command`, should we
+use the action from the ID or the command? Both seem reasonable.

```

Comment:
I'm a bit confused by what you're asking here. Could you provide an example in json maybe?

Reply:
Example provided :)